Software/formable generate load case import npz

Gbowker opened the pull request:
Please see issue #289 . This task is able to successfully import loadcases from a .npz file, pass it to a damask simulation, and it run successfully.

My build details:
```python
matflow-new==0.3.0a137
damask==3.0.0a7.post0
damask-parse==0.2.30
h5py==3.12.1
hpcflow-new2==0.2.0a187
formable==0.1.21
```

My example workflow `generate_load_case_import_npz.yml` which uses this task, as well as the file this task needs, `sampled_loadcase.npz` can be found in the .zip file attatched, should you want to test this yourself.
[generate_load_case_import_npz.zip](https://github.com/user-attachments/files/19346308/generate_load_case_import_npz.zip)

Before merging with develop this task really should be tested with every task which would use the loadcases after this one, however for now the only one that does use loadcases as far as I'm aware is the `simulate_ve_loading` task, which currently only is done using damask. So therefore this should be merged into develop soon as it is very useful for producing yield surface predictions or running many CP simulations with varying loads.

aplowman reviewed (changes requested):
Thanks @Gbowker. I think this should be probably implemented as a class method on [LoadCase](https://github.com/hpcflow/matflow-new/blob/49da0aec669e21da2950163316a6dd8a0ec377de/matflow/param_classes/load.py#L876), rather than a script, because it only uses Numpy. I would suggest naming it `from_npz_file`.

Could you also write a test that compares an input `data` dict with what you would expect the output to be? Perhaps @gcapes could assist you with this. It could also do with a more detailed doc string. For example, you are doing some processing in a `for` loop, what does this do? And what does the `idx` argument correspond to?

Gbowker commented:
Okay I've removed the script `data/scripts/read_FE_model_npz.py`, and just moved the function to `param_classes/load.py` but I'm certain I've done it wrong becuase all of the functions with the `@classmethod` decorator use `cls` and `steps` as inputs, but mine doesnt have this. 
I also need to change `data/template_components/task_schemas/` `method:import_npz` to point tio the param_class function now, as in [here](https://github.com/hpcflow/matflow-new/blob/158ac2e4cc47f848033f5e015a20aca5e9c551d5/matflow/data/template_components/task_schemas.yaml#L816)

How do I correctly implement this @aplowman ?

aplowman commented on matflow/data/template_components/task_schemas.yaml:
```diff
@@ -801,6 +801,23 @@
+      script_data_in: direct
+      script_data_out: direct
+      script_exe: python_script
+
```
No task schema is needed for a parameter class method, so this can be removed. You would invoke it like the other load case class methods (within the inputs block in the task where you need the load case). For example, instead of:

```yaml
- schema: simulate_VE_loading_damask
  inputs:
    load_case::uniaxial:
      ...
```
you would do:

```yaml
- schema: simulate_VE_loading_damask
  inputs:
    load_case::from_npz_file:
      ...
```

Gbowker commented:
Okay in commit [ec3225b](https://github.com/hpcflow/matflow-new/pull/334/commits/ec3225b382f8c1c32e01892cf56e4c200e42a9e2) I removed that objective from the task schema.

Commit [46682e1](https://github.com/hpcflow/matflow-new/pull/334/commits/46682e1c91be04082a5fd8823691b34cacb73406) was becuase I committed before pulling the automatic `pre-commit fixes` commit that github has done, so I had to pull that before pushing my commit.

How do I now change the `from_npz_file` function in `param_classes/load.py` so that it works correctly?

aplowman commented:
Take a look at the other `LoadCase` class methods for inspiration :) E.g. https://github.com/hpcflow/matflow-new/blob/f3ea5fa6363bfbe6bb756430e408aa1126579fca/matflow/param_classes/load.py#L944

Two things to note:

1. The first argument to a class method is the class itself, which conventionally is named `cls`, so you'll need to add that to your method.
2. The class method must return an instance of the class. So you'll need to have something like `return cls(...)` at the end of your method, where you pass in the correct arguments. In this case, that is a `steps` argument, which is a list of `LoadStep` objects. The `LoadStep` class is defined in this same file. So consult that class to understand what needs to be passed to it.

Gbowker commented:
I really need help here. @aplowman I've got it into the closest to your formatting as possible, but all of the other loadStep class methods have different inputs and outputs so i don't know how to proceed.

See my previous implementation as a script which worked the way I intended. Note how I sequence through indicies of loadcases of the file in the workflow, as each loadcase needs to be used for a different simulation - however the classmethods here appear to make only one loadcase. 

What I want to do explained in detail:
- I want to import a list of loadcases from a .npz file, each loadcase is for a different, unique simulation which I want to run, The index is what I sequence through. Think of it as a simulation index.
- Each loadcase though is a series of different loadcases, one for each increment of the simulation (don't ask me why - Mike did this not me)
- While I am making this a class method I need to still be able to sequence through them - since I am running the simulate_VE_loading_damask task repeatedly for each different loadcase index I extract from this npz file.
- However here in `param_classes/load.py` each of the class methods seem to represent ONE loadcase for ONE simulation. How do I get my class method to import a list of loadcases then only select one based on the index, which I want to sequence through in the workflow?
Does it work like this:
```yaml
- schema: simulate_VE_loading_damask
  inputs:
    loadcase::from_npz_file:
      sequence:
         index = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, ... , 99] # again, think of this like simulation index
```

aplowman commented:
> See my previous implementation as a script which worked the way I intended. Note how I sequence through indicies of loadcases of the file in the workflow, as each loadcase needs to be used for a different simulation - however the classmethods here appear to make only one loadcase.

Correct. The class methods are used for generating a single parameter, a load case, for a single simulation.

> * I want to import a list of loadcases from a .npz file, each loadcase is for a different, unique simulation which I want to run, The index is what I sequence through. Think of it as a simulation index.

Okay, so your simulation index is an input into the load case class method. You can use a sequence in your workflow to invoke the class method multiple times, each with a different simulation index.

> * Each loadcase though is a series of different loadcases, one for each increment of the simulation (don't ask me why - Mike did this not me)

In MatFlow terminology, a `LoadCase` is comprised of one or more `LoadStep`s. For a given simulation, each simulation increment can be associated with its own `LoadStep`. For an analagous example, you should look at the `LoadCase.uniaxial_cyclic` method and the corresponding `LoadStep.uniaxial_cyclic` method. This is more like your situation. So you want to modify your method `LoadStep.from_npz_file` such that it returns a *list* of `LoadStep` objects (in the same way `LoadStep.uniaxial_cyclic` does).

> * While I am making this a class method I need to still be able to sequence through them - since I am running the simulate_VE_loading_damask task repeatedly for each different loadcase index I extract from this npz file.
>
> * However here in `param_classes/load.py` each of the class methods seem to represent ONE loadcase for ONE simulation. How do I get my class method to import a list of loadcases then only select one based on the index, which I want to sequence through in the workflow?
>   Does it work like this:
> 
> ```yaml
> - schema: simulate_VE_loading_damask
>   inputs:
>     loadcase::from_npz_file:
>       sequence:
>          index = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, ... , 99] # again, think of this like simulation index
> ```

Yes, that's not a problem. You can set it up like this:

```yaml
  - schema: simulate_VE_loading_damask
    inputs:
      load_case::from_npz_file:
        npz_file_path: /path/to/npz/file
      homogenization: ...
      damask_phases: ...
      ...
    sequences:
      - path: inputs.load_case.idx
        values::from_range:
          start: 0 # inclusive
          stop: 100 # exclusive
          step: 1
```

Gbowker commented:
Ok this is working for me now! It at least for two simulations. @aplowman could you test this?
Thanks, attatched is the .npz I used. 
[sampled_loadcase.npz.zip](https://github.com/user-attachments/files/20296657/sampled_loadcase.npz.zip)

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,52 @@ def uniaxial_cyclic(
+        Parameters
+        ----------
+        npz_file_path
+            A string of the filepath to the loadcase.npz file.
```
Could you list which keys are required in this file, and what they represent here?

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,52 @@ def uniaxial_cyclic(
+                dt = inc_size[inc_size_idx]
+            dt = abs(dt) / strain_rate
+
+            load_cases.append(
```
```suggestion
            load_steps.append(
```

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,52 @@ def uniaxial_cyclic(
+                }
+            )
+
+        return [cls(**i)._remember_name_args(None, i) for i in load_cases]
```
```suggestion
        return [cls(**i)._remember_name_args(None, i) for i in load_steps]
```

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,52 @@ def uniaxial_cyclic(
+        u_sampled_split = data["u_sampled_split"]
+        strain_rate = data["strain_rate"]
+
+        load_cases = []
```
```suggestion
        load_steps = []
```

Gbowker commented:
> To that end, please could create a smaller version of your npz file (currently you have I think 100 indices in your test file; I would suggest cutting it down to say, three), and share this file, and also share what your expect this method to return from reading that file? I can then create a test in this PR.

Hi @aplowman , thanks for your review and help
[sampled_loadcase_2p.npz.zip](https://github.com/user-attachments/files/20506062/sampled_loadcase_2p.npz.zip)
Had to zip it to be able to send as .npz is not acceptable filetype here for some reason.

I expect this method to return all of the usual parameters that all the other loadcase class methods return. Those needed for a damask simulation:
```
    Boundary conditions for volume element loading.
    Parameters
    ----------
    total_time : float | int
        Total simulation time.
    num_increments
        Number of simulation increments.
    direction
        Direction or directions in which loading is done.
    normal_direction
        Direction of normal vector.
    target_def_grad : numpy.ma.core.MaskedArray, shape (3, 3)
        Deformation gradient aim tensor. Masked values correspond to unmasked values in
        `stress`.
    target_def_grad_rate : numpy.ma.core.MaskedArray, shape (3, 3)
        Deformation gradient rate tensor. Masked values correspond to unmasked values in
        `stress`.
    target_vel_grad : numpy.ma.core.MaskedArray, shape (3, 3)
        Velocity gradient aim tensor.
    stress : numpy.ma.core.MaskedArray, shape (3, 3)
        Stress tensor. Masked values correspond to unmasked values in
        `target_def_grad` or `target_def_grad_rate`.
    dump_frequency : int
        By default, 1, meaning results are written out every increment.
```

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,70 @@ def uniaxial_cyclic(
+            Filepath to the loadcase.npz file.
+            File contains dict of the following keys:
+            num_incs: 1D numpy array
+                Array of the total increments of each loadcase.
```
```suggestion
                Array of the total number of increments to use for each loadcase.
```

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,70 @@ def uniaxial_cyclic(
+            inc_size: 2D numpy array
+                Array of the amount of strain each loadstep of each loadcase should undergo in the damask simulation
+            inc_size_final: 2D numpy array
+                Array of amount of strain of final loadsteps
```
What does the second dimension represent for these two (assuming the first is corresponding to the load case that we extract using `idx`)?

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,70 @@ def uniaxial_cyclic(
+                Array of the amount of strain each loadstep of each loadcase should undergo in the damask simulation
+            inc_size_final: 2D numpy array
+                Array of amount of strain of final loadsteps
+            u_sampled_split: 4D numpy array
+                Array of strain matricies (3x3) for each loadstep (1st,2nd D) should undergo for each loadcase (3nd D),
+                for each simulation (4th D). Sampled from the elements of an FE model.
```
Isn't the load*case* the first dimension, the load step the second dimension, and the strain matrix the final dimensions?

Typo: `matrices`

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,70 @@ def uniaxial_cyclic(
+                for each simulation (4th D). Sampled from the elements of an FE model.
+            LE_eval_sampled_split: 3D numpy array
+                Array of principle component vectors of strain matricies (3x1), for each loadstep (1st D),
+                should undergo for each loadcase (2nd D), for each simulation (3rd D).
```
This doesn't appear to be used, so I would remove this for simplicity.

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,70 @@ def uniaxial_cyclic(
+        Parameters
+        ----------
+        npz_file_path: str
+            Filepath to the loadcase.npz file.
+            File contains dict of the following keys:
```
```suggestion
            Filepath to the npz file, which must be dict-like with at least the following keys:
```

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,70 @@ def uniaxial_cyclic(
+                Array of principle component vectors of strain matricies (3x1), for each loadstep (1st D),
+                should undergo for each loadcase (2nd D), for each simulation (3rd D).
+            strain_rate: 1D numpy array of one float
+                Array of strain rate to be used for every simulation
```
This looks like a scalar array from your example data, i.e. `np.array(0.01)`. To avoid confusion, I would not refer to it as an array:
```suggestion
                Scalar strain rate to be used for every simulation
```

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,70 @@ def uniaxial_cyclic(
+                should undergo for each loadcase (2nd D), for each simulation (3rd D).
+            strain_rate: 1D numpy array of one float
+                Array of strain rate to be used for every simulation
+            random_sample_LE: 1D numpy array
+                Array of principle component of strain in one direction.
```
See comment for `LE_eval_sampled_split` above.

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,70 @@ def uniaxial_cyclic(
+        idx,
+    ) -> list[Self]:
+        """
+        Read loading data from a numpy .npz file and format it appropriately.
+        Designed for running large arrays of simulations which each use loadcase of the
+        corresponding index.
```
```suggestion
        Construct a list of load steps using data from a Numpy .npz file. This is designed for running large arrays of simulations from the data in this file, where each uses a loadcase specified by a given index (`idx`).
```

aplowman commented on matflow/param_classes/load.py:
```diff
@@ -871,6 +871,70 @@ def uniaxial_cyclic(
 
+    @classmethod
+    def from_npz_file(
+        cls,
+        npz_file_path,
+        idx,
```
```suggestion
        cls,
        npz_file_path: str,
        idx: int,
```

Gbowker commented:
🥳
